### ISSUE TITLE
[Modular CLI]: crash during runtime

### ISSUE THREAD

**Issue by pmatos** [modular/mojo#1369]
### Issue description

My code is crashing - and unfortunately the backtrace is non-existent. I imagine the mojo cli is distributed without debugging syms. 

### Steps to reproduce

1. Unzip 
[aoc-2.mojo.gz](https://github.com/modularml/mojo/files/13545283/aoc-2.mojo.gz)
1.1 Reduced version:
```
fn main():
    let s: String = "hello"
    var i: Int = 0
    while s[i]:
        i += 1
```
2. Run:
```
$ /home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo /home/pmatos/dev/aoc2023/aoc-2.mojo
Please submit a bug report to https://github.com/modularml/mojo/issues and include the crash backtrace along with all the relevant source codes.
Stack dump:
0.      Program arguments: /home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo /home/pmatos/dev/aoc2023/aoc-2.mojo
#0 0x000056052ec1fb97 (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x5c7b97)
#1 0x000056052ec1d76e (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x5c576e)
#2 0x000056052ec2026f (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x5c826f)
#3 0x00007f115e45c710 (/usr/lib/libc.so.6+0x3e710)
#4 0x00007f107c001099 
[1]    45455 segmentation fault (core dumped)  /home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo
```

### Version Info

```shell
- Provide Modular CLI version by pasting the output of `modular -v`
mojo 0.6.0 (d55c0025)
- What OS did you install modular CLI on ?
Archlinux AUR mojo-git
```

**Comment by pmatos** [comment]
Unknown if it's related but while reducing mojo using cvise, I got the following reduced program:
```
fn is_valid_game()  :
    let Bool = True
            False
```

Running it:
```
mojo aoc-2.mojo
Please submit a bug report to https://github.com/modularml/mojo/issues and include the crash backtrace along with all the relevant source codes.
Stack dump:
0.	Program arguments: mojo aoc-2.mojo
1.	Crash resolving decl body at loc("/home/pmatos/dev/aoc2023/aoc-2.mojo":1:1)
    >> fn is_valid_game()  :
       ^....................
    >>     let Bool = True
    >>             False
       .................<
2.	Crash parsing statement at loc("/home/pmatos/dev/aoc2023/aoc-2.mojo":3:13)
    >>             False
                   ^....<
 #0 0x0000557b28599b97 (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x5c7b97)
 #1 0x0000557b2859776e (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x5c576e)
 #2 0x0000557b2859a26f (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x5c826f)
 #3 0x00007fc5aaa5c710 (/usr/lib/libc.so.6+0x3e710)
 #4 0x0000557b28b34964 (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0xb62964)
 #5 0x0000557b29c61855 (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x1c8f855)
 #6 0x0000557b289d1763 (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x9ff763)
 #7 0x0000557b28a3689e (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0xa6489e)
 #8 0x0000557b28a2e359 (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0xa5c359)
 #9 0x0000557b28a633ee (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0xa913ee)
#10 0x0000557b28a5d9d4 (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0xa8b9d4)
#11 0x0000557b28a5d886 (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0xa8b886)
#12 0x0000557b28a0c81c (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0xa3a81c)
#13 0x0000557b28a032e5 (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0xa312e5)
#14 0x0000557b28a014c9 (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0xa2f4c9)
#15 0x0000557b289c247d (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x9f047d)
#16 0x0000557b289c2f0b (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x9f0f0b)
#17 0x0000557b285632ff (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x5912ff)
#18 0x0000557b2854fe9b (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x57de9b)
#19 0x0000557b285621d8 (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x5901d8)
#20 0x0000557b2854948d (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x57748d)
#21 0x00007fc5aaa45cd0 (/usr/lib/libc.so.6+0x27cd0)
#22 0x00007fc5aaa45d8a __libc_start_main (/usr/lib/libc.so.6+0x27d8a)
#23 0x0000557b28548a1e (/home/pmatos/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x576a1e)
[1]    1076016 segmentation fault (core dumped)  mojo aoc-2.mojo
```

**Comment by rarebreed** [comment]
The line `let Bool = True` isn't valid.  Try this instead:

```
fn is_valid_game():
    let b: Bool = True
    False  # this does nothing.  mojo is not an expression oriented language and do an implicit return
```

**Comment by rarebreed** [comment]
> @rarebreed I don't see how if `let Bool = True` invalid. It just bind `True` to a variable named `Bool` that happens to shadow the builtin `Bool`. The following runs without problem:
> 
> ```
> fn a():
>   let Bool = False
>   print(Bool)
> 
> fn b():
>   let Int = 1
>   False
> 
> fn c():
>   let Bool = False
>   1
> ```
> 
> However, this doesn't:
> 
> ```
> fn main():
>   let Bool = False
>   False
> ```

Interesting...so the builtin types aren't keywords?  I had assumed it was invalid since the code crashed my parser, but just putting that on the line by itself does indeed work.

But my spidey sense feels like that's a bad idea.  That will make this code wrong:

```
fn test():
    let Bool = True
    let b: Bool = False
```

I can easily see a programmer forgetting to do `let b: Bool = True`, then scratch their heads why the rest of the code doesn't allow the type declaration.

**Comment by rarebreed** [comment]
Since mojo doesn't allow shadowing like rust does, I think it will be odd to allow variables to be named the same as a type.  Java for example won't let you do `String String = "String;".  It is true that rust will let you do it, though I'd raise my eyebrows in a code review.  It's also an unfortunate part of python, where people commonly shadow builtin modules or functions (eg `id` is a common one).

The example I gave was pretty trivial, but you could also have code like this

```
fn closurer() -> fn() capturing -> Bool:
    var Bool = True
    fn inner() -> Bool:
        return Bool
    return inner
```
yields this error

```
/home/stoner/Projects/excursor/excursor/mojo/global.mojo:7:19: error: cannot use a dynamic value in type specification
    fn inner() -> Bool:
                  ^~~~
mojo: error: failed to parse the provided Mojo
```

**Comment by rarebreed** [comment]
So I guess the question is, if mojo doesn't allow renaming/shadowing of a variable (which I would imagine is just a symbol in a namespace table per scope), should the builtin types, functions, etc also be included into the namespace as symbols to prevent shadowing?

Alternatively, I suppose mojo could do some kind of analysis to know that `Bool` in the return type must be a type, and not a value.  Except that python allows for `Literal` types.  So this is valid python

```
from typing import Literal

def always_true(): -> Literal[True]:
    return True
```

So maybe it might be hard to disambiguate a return type of `Bool` versus `True`.  On one hand, I feel like renaming/shadowing a builtin is usually a bad idea (even in rust, shadowing was controversial).  However, there might be some valid use cases.

**Comment by pmatos** [comment]
> The line `let Bool = True` isn't valid. Try this instead:
> 
> ```
> fn is_valid_game():
>     let b: Bool = True
>     False  # this does nothing.  mojo is not an expression oriented language and do an implicit return
> ```
> 
> It would be nice if the VS Code mojo analyzer could spot this before running/compiling (the original code crashes the parser)

Still crashing in 0.6, although the original program now returns:

```
$ mojo aoc-2.mojo
/home/pmatos/dev/aoc2023/aoc-2.mojo:52:53: error: inout arguments may not have defaults
fn is_prefix(prf: String, s: String, inout idx: Int = 0) -> Bool:
                                                    ^ ~
mojo: error: failed to parse the provided Mojo
```

which is fine.

**Comment by pmatos** [comment]
I feel tempted to close this - as fixed in 0.6, and open an issue for the 

```
fn is_valid_game()  :
    let Bool = True
            False
```

case which still crashes.

**Comment by pmatos** [comment]
Ah wait - I found that the original problem still persists and reduced it to:
```
fn main():
    let s: String = "hello"
    var i: Int = 0
    while s[i]:
        i += 1

```

This is clearly due to a bug in my code, but it still feels strange to have mojo segfault.

**Comment by laszlokindrat** [comment]
> I feel tempted to close this - as fixed in 0.6, and open an issue for the
```mojo
fn is_valid_game()  :
    let Bool = True
            False
```
> case which still crashes. wdyt?

Could you please open a separate issue for this? Seems like an unrelated problem with the parser.
